Add interruptible key agreement negative tests

paul-elliott-arm opened the pull request:
## Description

Negative tests for Interruptible Key Agreement.

Resolves #9109 

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** ~~provided |~~ not required because: Part of interruptible ECC work
- [x] **development PR** ~~provided # |~~ not required because: This is the development PR 
- [x] **framework PR** provided Mbed-TLS/mbedtls-framework#82 ~~not required~~
- [x] **TF-PSA-Crypto PR** provided Mbed-TLS/tf-psa-crypto#120 ~~not required~~ 
- [x] **3.6 PR** ~~provided # |~~ not required because: New feature for 4.0 
- [x] **2.28 PR** ~~provided # |~~ not required because: New feature for 4.0 
- **tests**  provided ~~| not required because:~~

waleed-elmelegy-arm reviewed (changes requested):
Thanks for the changes.

Some missing tests for interruptible operations state testing:
* Testing that calling complete() without calling setup() successfully results in an error.
* Testing that calling complete() again after finishing results in an error.
* Testing calling setup 2 times in a row results in an error.
* Testing that calling abort after an error resets the state and causes setup() to be successful.
* Testing that when passing a non ECC key result in an error that indicates it's not supported.

Apart from that we should add an iop key agreement initialisation test similar to this [one](https://github.com/Mbed-TLS/mbedtls/pull/9639/files#diff-669d3446e5ada9b83db697969e8f90b82704f19ea8d4edfef1b8be281a0f68a1R10167).

paul-elliott-arm commented:
> > All done bar this, which won't result in an error - could you elaborate more on what you are expecting here? There is no finished state per se in the code as it stands.
> 
> As I understand after a complete operation has finished either because of success or an error the operation should not be usable (internal abort() should be called) and it should be reset by calling setup() in order to be usable again. Ideally calling abort should be done by the user but we agreed to do it ourselves to protect the user from misuse.

As discussed, this wasn't the case, fixed it here, and added tests.

gilles-peskine-arm commented:
> As I understand after a complete operation has finished either because of success or an error the operation should not be usable (internal abort() should be called) and it should be reset by calling setup() in order to be usable again. Ideally calling abort should be done by the user but we agreed to do it ourselves to protect the user from misuse.

Actually there's an additional subtlety here. Indeed, after complete() returns anything but `OPERATION_INCOMPLETE`, the operation can't be used, except for one thing: you can call `get_num_ops` to know how long the operation took. So a complete() operation shouldn't do a full abort: it should clean out all cryptographic data, but leave the op count. `setup()` needs to reset the op count to 0.

There's a state diagram in https://github.com/user-attachments/files/15927984/IHI0086-PSA_Certified_Crypto_API-1.3.0-interruptible-operations-draft3.pdf figure 2 page 31, but it doesn't capture this. In the inactive state, complete() is not valid, but get_num_ops() can return nonzero.

paul-elliott-arm commented:
> Actually there's an additional subtlety here. Indeed, after complete() returns anything but `OPERATION_INCOMPLETE`, the operation can't be used, except for one thing: you can call `get_num_ops` to know how long the operation took. So a complete() operation shouldn't do a full abort: it should clean out all cryptographic data, but leave the op count. `setup()` needs to reset the op count to 0.

I believe after this PR the code is currently correct, as this stands. The number of ops is not wiped until the user calls the abort() method.

waleed-elmelegy-arm reviewed (approved):
LGTM, Thanks.

davidhorstmann-arm reviewed (approved):
LGTM (but sadly needs moving to TF-PSA-Crypto)

waleed-elmelegy-arm reviewed (dismissed):
LGTM.

yanesca reviewed (approved):
Did the split and arrived at the same content.

paul-elliott-arm commented:
Apologies had to rebase again, due to code_style.py framework issues. Updated both submodule pointers to their requisite HEADs.

gilles-peskine-arm reviewed (approved):
Approved at b08dc92aaaecf24b2178df412c13960d2803bd9f based on:

* [@yanesca's approval](https://github.com/Mbed-TLS/mbedtls/pull/9735#pullrequestreview-2512256399) of 3cc74803f21e5e0750ef4cbdee6e40f1b8bff451
* From, there has only been a correct rebase, and submodule updates.
* `tf-psa-crypto` is updated to the head of `development`, which is always ok.
* `framework` is aligned with `tf-psa-crypto/framework`, and which already contains https://github.com/Mbed-TLS/mbedtls-framework/pull/82 which is needed here.
* `framework` is moving forward.
